### PR TITLE
Update Readme

### PR DESCRIPTION
**Pull request by AdithyaKrishnan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,592 @@
...
+
+- The verification includes checking Boot Loader, TEE (Trusted Execution Environment), SNP (Secure Nested Paging), Microcode, and Hardware ID values.
+
+- The function also allows for separate verification of the TCB (Trusted Computing Base) and the digital signature.
+
+- Verification results are printed to the console.
+
+6. **Helper Functions**:
+
+- Several helper functions are defined within the modules to facilitate various tasks, including locating certificates in a directory, comparing certificate values, and verifying attestation report signatures.
```
Comment (DGonzalezVillal) [inline README.md:288]: Although this is a good summary of how the code works and I appreciate the effort, I don't think this is needed in the README. The README should focus on how to use the tool, not so much on how it works. Unless you disagree @larrydewey @tylerfanelli

File: README.md [hunk 2]
```diff
@@ -1,8 +1,592 @@
...
+- `-h, --help`: Show a help message.
+
+Example usage:
+```bash
+# Displaying a derived key
+snpguest display key derived-key.bin
+```
+---
+
+# SNP Examples
```
Comment (DGonzalezVillal) [inline README.md:464]: Your examples are just a repeat from what you have above in commands. I would maybe use this section to show how would you go about going through a regular attestation workflow and an extended attestation workflow. This in addition to the flowcharts you added at the bottom would be very helpful.

File: README.md [hunk 3]
```diff
@@ -1,8 +1,592 @@
...
+snpguest verify attestation ./certs attestation-report.bin
+# Verify Attestation Signature only
+snpguest verify attestation ./certs attestation-report.bin --signature
+```
+
+---
+## Extended Attestation Flowchart
+![alt text](https://github.com/virtee/snpguest/blob/main/docs/extended.png?raw=true)
+## Regular Attestation Flowchart
+![alt text](https://github.com/virtee/snpguest/blob/main/docs/regular.png?raw=true)
```
Comment (DGonzalezVillal) [inline README.md:588]: I'm not sure if this is how you attach images. At least for me, in you repository, I can't see the images in the README.

File: README.md
Comment (DGonzalezVillal) [inline README.md:1]: In addition to the other comments I left, I would also add a section on how to compile the tool. You also fail to mention the whole hyper-v section.

**Review by DGonzalezVillal — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -1,8 +1,592 @@
...
+
+## Summary
+
+The "src" directory in the `snpguest` GitHub repository contains several Rust source code files that collectively implement the functionality of the SNP Guest utility. Here's a summary of the functionality of each of these files:
+
+1. **`certs.rs`**:
+
+- This file contains code related to managing certificates.
+
+- It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD PSP.
```
Comment (larrydewey) [inline README.md:32]:
```suggestion
- It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD Secure Processor.
```

File: README.md [hunk 5]
```diff
@@ -1,8 +1,592 @@
...
+
+- It contains subcommands for displaying attestation reports and derived keys.
+
+- Submodules `report_display` and `key_display` handle the respective display functionalities.
+
+3. **`fetch.rs`**:
+
+- This file defines the CLI for fetching certificates.
+
+- It contains subcommands for fetching various types of certificates from the AMD PSP.
```
Comment (larrydewey) [inline README.md:46]:
```suggestion
- It contains subcommands for fetching various types of certificates from the AMD Secure Processor.
```

File: README.md [hunk 6]
```diff
@@ -1,8 +1,592 @@
...
+
+- This file contains code related to managing certificates.
+
+- It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD PSP.
+
+2. **`display.rs`**:
+
+- This file defines the CLI for displaying attestation reports and derived keys.
+
+- It contains subcommands for displaying attestation reports and derived keys.
```
Comment (larrydewey) [inline README.md:38]: I think it would be beneficial to actually list out each sub-command, and what they are expected to do and how they are expected to be used.

Comment (larrydewey) [inline README.md:38]: I think it would be beneficial to actually list out each sub-command, and what they are expected to do and how they are expected to be used.

File: README.md [hunk 7]
```diff
@@ -1,8 +1,592 @@
...
+
+- The main program defines a CLI interface using the `structopt` crate.
+
+- It has subcommands for different operations, including reporting, certificate management, fetching, verification, displaying, and key generation.
+
+- The `quiet` flag is provided to suppress console output if set.
+
+4. **Main Function** (`main`):
+
+- The main function initializes the `env_logger` for logging.
```
Comment (larrydewey) [inline README.md:113]: It would be good to explain how to set the environment variable to change the log level at run-time.

File: README.md (hunk 3 above)
Comment (larrydewey) [reply README.md:588]: I cannot see them, either

File: README.md (hunk 1 above)
Comment (larrydewey) [reply README.md:288]: I agree

**Review by larrydewey — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (tylerfanelli) [reply README.md:288]: I agree as well. However, much of the code comments would fit nicely in their respective source code files.

File: README.md [hunk 8]
```diff
@@ -1,8 +1,592 @@
-# snpguest
+**SNPguest** is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key generation, and more.
```
Comment (tylerfanelli) [inline README.md:1]:
```suggestion
**SNPguest** is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key fetching, and more.
```

I don't think there is any key generation in this tool, moreso fetching keys from the KDS that already exist.

File: README.md (hunk 4 above)
Comment (tylerfanelli) [reply README.md:32]: Just for my reference, what's the difference between the two? I usually use them interchangeably. Or is this change because some users may not know what PSP stands for?

File: README.md (hunk 6 above)
Comment (tylerfanelli) [reply README.md:38]: Right, the [sevctl README](https://github.com/virtee/sevctl/blob/main/README.md) does just this, and can be used for reference.

File: README.md (hunk 1 above)
Comment (tylerfanelli) [reply README.md:288]: Much of this would also be useful in a man page as well.

File: README.md [hunk 9]
```diff
@@ -1,8 +1,592 @@
...
+
+```bash
+snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
+```
+
+Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request, and for Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. If the `--random` flag is used, the data will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1.
+
+Options:
+- `-h, --help`: Show a help message.
+- `-r, --random`: Generate 64 random bytes of data for the report request.
```
Comment (tylerfanelli) [inline README.md:308]:
```suggestion
- `-r, --random`: Generate 64 random bytes of data for the report request. (Not used in Hyper-V)
```

File: README.md [hunk 10]
```diff
@@ -1,8 +1,592 @@
...
+
+### Commands
+
+#### snpguest report
+
+```bash
+snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
+```
+
+Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request, and for Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. If the `--random` flag is used, the data will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1.
```
Comment (tylerfanelli) [inline README.md:304]:
```suggestion
Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request.

For Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. The data already generated from the hypervisor will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1. `--random` is not available in Hyper-V, as the request data is pre-generated.
```

File: README.md [hunk 11]
```diff
@@ -1,8 +1,592 @@
...
+Example usage:
+```bash
+# Verifying Certs
+snpguest verify certs ./certs
+```
+
+#### snpguest guest verify attestation
+
+```bash
+snpguest verify attestation $CERTS_DIR $ATT_REPORT_PATH [-t, --tcb] [-s, --signature]
```
Comment (tylerfanelli) [inline README.md:392]: Why is `[-s, --signature]` an option? Shouldn't the signature be a part of the attestation report?

@DGonzalezVillal @larrydewey as in, must there be a separate flag for the signature?

**Review by tylerfanelli — CHANGES_REQUESTED** [review]
Much of this fits into a README, but a lot could also fit into a man page. Code-specific descriptions are better left to source code comments.

However, this is very descriptive and thorough about the usage of the tool (a huge plus for those looking to figure out how the tools works).

Great diagrams as well!

File: README.md (hunk 11 above)
Comment (DGonzalezVillal) [reply README.md:392]: Hey, this allows you to perform attestation on only the signature part or only the tcb part.

File: README.md [hunk 12]
```diff
@@ -1,8 +1,574 @@
...
+
+1. **Command-Line Interface (CLI) Definition**: The code defines a CLI with two subcommands:
+
+- `CA`: Fetches the certificate authority chain (ARK & ASK) from the KDS.
+
+- `Vcek`: Fetches the VCEK (Virtualization Code Execution Key) from the KDS.
+
+2. **Enumeration and Parsing**: It defines an enumeration called `ProcType` to represent processor types ("Milan" or "Genoa"). It also implements the `FromStr` trait for parsing processor types and the `fmt::Display` trait for displaying them.
+
+3. **Command Handling**: The `cmd` function is responsible for handling the selected subcommand and performing the corresponding action. It calls either the `cert_authority::fetch_ca` function or the `vcek::fetch_vcek` function based on the selected subcommand.
```
Comment (DGonzalezVillal) [inline README.md:188]: Same thing as above. Very good summary, but not part of a README

File: README.md [hunk 13]
```diff
@@ -1,8 +1,574 @@
...
+
+```bash
+snpguest verify certs $CERTS_DIR
+```
+
+**Step 5.**
+
+```bash
+snpguest verify attestation $CERTS_DIR $ATT_REPORT_PATH [-t, --tcb] [-s, --signature]
+```
```
Comment (DGonzalezVillal) [inline README.md:247]: Explain what each step is doing.

File: README.md [hunk 14]
```diff
@@ -1,8 +1,574 @@
...
+Options:
+- `-h, --help`: Show a help message.
+
+Example usage:
+```bash
+# Requesting VCEK
+snpguest fetch vcek der milan ./certs-kds attestation-report.bin
+```
+
+#### snpguest guest verify certs
```
Comment (DGonzalezVillal) [inline README.md:329]:
```suggestion
#### snpguest verify certs
```

File: README.md [hunk 15]
```diff
@@ -1,8 +1,574 @@
...
+Options:
+- `-h, --help`: Show a help message.
+
+Example usage:
+```bash
+# Verifying Certs
+snpguest verify certs ./certs
+```
+
+#### snpguest guest verify attestation
```
Comment (DGonzalezVillal) [inline README.md:346]:
```suggestion
#### snpguest guest attestation
```

File: README.md [hunk 16]
```diff
@@ -1,8 +1,574 @@
...
+
+Options:
+- `-h, --help`: Show a help message.
+
+Example usage:
+```bash
+# Displaying a derived key
+snpguest display key derived-key.bin
+```
+---
```
Comment (DGonzalezVillal) [inline README.md:419]: This should be the first part of the README. The way I would divide would be:
A section for the main command, and then a small explanation of the subcommands. 
For Example:
```
### verify
Command to do different types of verification
#### verify certs

snpguest verify certs $CERTS_DIR


Verifies that the provided certificate chain has been properly signed by each certificate. The user needs to provide a directory where all three certificates (ARK, ASK, and VCEK) are stored. An error will be raised if any of the certificates fail verification.

Options:
- `-h, --help`: Show a help message.

Example usage:
# Verifying Certs
snpguest verify certs ./certs
```

Or something along those lines. Again check SEVCTL README

File: README.md [hunk 17]
```diff
@@ -1,8 +1,574 @@
...
+- `-s, --signature`: Just verify the signature of the attestation report.
+
+Example:
+
+```bash
+# Verify Attestation
+snpguest verify attestation ./certs attestation-report.bin
+# Verify Attestation Signature only
+snpguest verify attestation ./certs attestation-report.bin --signature
+```
```
Comment (DGonzalezVillal) [inline README.md:539]: This seems very repetitive to what you have in the previous section. The examples should be way shorter. What I would do is go back to the attestation workflow examples and just explain what you're doing in each step. Maybe then add an example for the derived key.

**Review by DGonzalezVillal — CHANGES_REQUESTED** [review]
It's still very disorganized and a lot of unnecessary information for a README. I would go back and see other README examples in the repository. I would recommend to remove the summary and file structure sections. A lot of that information I would put as source code documentation in the form of comments. Then the commands section needs restructuring, refer to my comments to see what I mean. You have 2 example sections, one isn't very clear and the other is too wordy. Combine those 2 to show people how to perform both regular & extended attestation in the tool. The building part looks good, you should still mention compiling for hyper-v.

File: README.md [hunk 18]
```diff
@@ -1,8 +1,574 @@
-# snpguest
+`snpguest` is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key fetching, and more.
```
[the remark this answers is not in the document]
Comment (AdithyaKrishnan) [reply README.md:1]: I believe @tylerfanelli had suggested to use the term fetch instead of derivation. Can we come to a consensus?

[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply README.md:1]: Derived key fetching

File: README.md [hunk 19]
```diff
@@ -1,8 +1,376 @@
...
+## Usage
+### 1. `help`
+
+Every `snpguest` (sub)command comes with a `--help` option for a description on its use. 
+
+Usage
+```console
+snpguest --help
+```
+Apart from `help`, there are 6 other subcommands for `snpguest` that have been listed below with their purpose.
```
Comment (DGonzalezVillal) [inline README.md:32]:
```suggestion

```

File: README.md [hunk 20]
```diff
@@ -1,8 +1,376 @@
...
+
+    - `$KEY_PATH` : Specifies the path of the stored derived key to display.
+
+    Example
+    ```bash
+    snpguest display key derived-key.bin
+    ```
+### 4. `fetch`
+
+Command to Requests certificates.
```
Comment (DGonzalezVillal) [inline README.md:99]:
```suggestion
Command to Requests certificates from the KDS.
```

File: README.md [hunk 21]
```diff
@@ -1,8 +1,376 @@
...
+
+Usage
+```bash
+snpguest fetch <SUBCOMMAND>
+```
+
+Subcommands
+1. `ca`
+
+    Requests the certificate authority chain (ARK & ASK) from the KDS. The user needs to specify the certificate encoding to store the certificates in (PEM or DER). Currently, only PEM and DER encodings are supported. Both certificates will be in the same encoding. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the certificates will be stored. If the certificates already exist in the provided directory, they will be overwritten.
```
Comment (DGonzalezVillal) [inline README.md:109]:
```suggestion
    Requests the certificate authority chain (ARK & ASK) from the KDS. The user needs to specify the certificate encoding to store the certificates in (PEM or DER). Currently, only PEM and DER encodings are supported. Both certificates will be in the same encoding. The user must specify their host processor model. The user also needs to provide the path to the directory where the certificates will be stored. If the certificates already exist in the provided directory, they will be overwritten.
```

File: README.md [hunk 22]
```diff
@@ -1,8 +1,376 @@
...
+
+    Usage
+    ```bash
+    snpguest fetch ca $ENCODING $PROCESSOR_MODEL $CERTS_DIR
+    ```
+    Arguments
+    
+    - `$ENCODING` : Specifies the certificate encoding to store the certificates in (PEM or DER)
+
+    - `$PROCESSOR_MODEL` : Specifies the host processor model (Milan or Genoa).
```
Comment (DGonzalezVillal) [inline README.md:119]:
```suggestion
    - `$PROCESSOR_MODEL` : Specifies the host processor model. 
```

File: README.md [hunk 23]
```diff
@@ -1,8 +1,376 @@
...
+    - `$CERTS_DIR` : Specifies the directory to store the certificates in.
+
+    Example
+    ```bash
+    snpguest fetch ca der milan ./certs-kds
+    ```
+
+2. `vcek`
+
+    Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.
```
Comment (DGonzalezVillal) [inline README.md:130]:
```suggestion
    Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model. The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.
```

File: README.md [hunk 24]
```diff
@@ -1,8 +1,376 @@
...
+
+    Usage
+    ```bash
+    snpguest fetch vcek $ENCODING $PROCESSOR_MODEL $CERTS_DIR $ATT_REPORT_PATH
+    ```
+    Arguments
+    
+    - `$ENCODING` : Specifies the certificate encoding to store the certificates in (PEM or DER)
+
+    - `$PROCESSOR_MODEL` : Specifies the host processor model (Milan or Genoa).
```
Comment (DGonzalezVillal) [inline README.md:140]:
```suggestion
    - `$PROCESSOR_MODEL` : Specifies the host processor model. 
```

File: README.md [hunk 25]
```diff
@@ -1,8 +1,376 @@
...
+    - `$ATT_REPORT_PATH` : Specifies the path of the stored attestation report.
+
+    Example
+    ```bash
+    snpguest fetch vcek der milan ./certs-kds attestation-report.bin
+    ```
+
+### 5. `key` 
+
+Creates the derived key based on input parameters and stores it. `$KEY_PATH` is the path to store the derived key. `$ROOT_KEY_SELECT` is the root key from which to derive the key (either "vcek" or "vmrk"). The `--guest_field_select` option specifies which Guest Field Select bits to enable as a 6-digit binary string. For each bit, 0 denotes off, and 1 denotes on (e.g., `--guest_field_select 100001` denotes Guest Policy:On, Image ID:Off, Family ID:Off, Measurement:Off, SVN:Off, TCB Version:On). The `--guest_svn` option specifies the guest SVN to mix into the key, and the `--tcb_version` option specifies the TCB version to mix into the derived key. The `--vmpl` option specifies the VMPL level the Guest is running on and defaults to 1.
```
Comment (DGonzalezVillal) [inline README.md:153]: On guest field select explanation, instead of showing example, just explain what each bit is. User can see example underneath.

File: README.md [hunk 26]
```diff
@@ -1,8 +1,376 @@
...
+snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
+```
+
+Arguments
+
+- `$ATT_REPORT_PATH` : Specifies the path where the attestation report would be stored.
+
+- `$REQUEST_FILE` : File where the data generated from the hypervisor will be written into.
+
+Options
```
Comment (DGonzalezVillal) [inline README.md:211]: Missing platform flag

**Review by DGonzalezVillal — CHANGES_REQUESTED** [review]
Looks great! Only a few last comments and it should be good!

File: README.md [hunk 27]
```diff
@@ -1,8 +1,378 @@
...
+
+- `-r, --random`: Generate 64 random bytes of data for the report request. (Not used in Hyper-V)
+- `-p, --platform` : Use platform provided 64 bytes of data for the request report.
```
Comment (larrydewey) [inline README.md:216]:
```suggestion
- `-p, --platform` : Use platform provided 64 bytes of data for the request report (only available for Hyper-V).
```

File: README.md [hunk 28]
```diff
@@ -1,8 +1,378 @@
...
+
+Arguments
+
+- `$ATT_REPORT_PATH` : Specifies the path where the attestation report would be stored.
+
+- `$REQUEST_FILE` : File where the data generated from the hypervisor will be written into.
+
+Options
+
+- `-r, --random`: Generate 64 random bytes of data for the report request. (Not used in Hyper-V)
```
Comment (larrydewey) [inline README.md:215]:
```suggestion
- `-r, --random`: Generate 64 random bytes of data for the report request (not available for Hyper-V).
```

File: docs/snpguest.1.adoc [hunk 29]
```diff
@@ -41,6 +41,7 @@ COMMANDS
     
     options:
     -h, --help  show a help message
+    -p, --platform Use platform generated 64 bytes of data to pass for the report request
```
Comment (larrydewey) [inline docs/snpguest.1.adoc:44]:
```suggestion
    -p, --platform Use platform generated 64 bytes of data to pass for the report request (only available for Hyper-V)
```

File: docs/snpguest.1.adoc [hunk 30]
```diff
@@ -41,6 +41,7 @@ COMMANDS
...
     -r, --random  Generate 64 random bytes of data to pass for the report request
```
Comment (larrydewey) [inline docs/snpguest.1.adoc:45]:
```suggestion
    -r, --random  Generate 64 random bytes of data to pass for the report request (not available for Hyper-V)
```

File: src/certs.rs [hunk 31]
```diff
@@ -1,4 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
+//This file contains code related to managing certificates. It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD Secure Processor.
```
Comment (larrydewey) [inline src/certs.rs:2]:
```suggestion
// This file contains code related to managing certificates. It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD Secure Processor.
```

File: src/key.rs [hunk 32]
```diff
@@ -1,4 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
+// - This file contains code for fetching derived keys from root keys. It also includes functions for requesting and saving derived keys.
```
Comment (larrydewey) [inline src/key.rs:2]:
```suggestion
// This file contains code for fetching derived keys from root keys. It also includes functions for requesting and saving derived keys.
```

**Review by larrydewey — CHANGES_REQUESTED** [review]
Just a couple minor adjustments. With these, i will approve it.

**Review by larrydewey — APPROVED** [review]
LGTM!

**Review by DGonzalezVillal — APPROVED** [review]
Awesome, looks great to me. Good job @AdithyaKrishnan
